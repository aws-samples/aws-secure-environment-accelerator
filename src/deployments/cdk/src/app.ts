import * as cdk from '@aws-cdk/core';
import { PhaseDeploy } from './apps/shared';
import { AccountStacks } from './common/account-stacks';
import { loadAccounts, getAccountId } from './utils/accounts';
import { loadAcceleratorConfig } from './utils/config';
import { loadContext } from './utils/context';
import { loadStackOutputs } from './utils/outputs';
import { loadLimits, Limiter } from './utils/limits';

export interface PhaseInfo {
  readonly runner: () => Promise<PhaseDeploy>;
  readonly name: string;
  readonly id: string;
}

// Right now there are only phases -1, 0, 1, 2, 3, 4, 5
export const phases: PhaseInfo[] = [-1, 0, 1, 2, 3, 4, 5].map(id => ({
  runner: () => import(`./apps/phase-${id}`).then(phase => phase.deploy),
  id: `${id}`,
  name: `${id}`,
}));

export interface AppProps {
  phaseId: string;
  region?: string;
  accountKey?: string;
  useTempOutputDir?: boolean;
}

export async function deploy(props: AppProps): Promise<cdk.Stage[]> {
  const { accountKey, phaseId, region, useTempOutputDir } = props;
  const phase = phases.find(p => p.id === phaseId);
  if (!phase) {
    throw new Error(`Cannot find phase with ID ${phaseId}`);
  }

  const acceleratorConfig = await loadAcceleratorConfig();
  const accounts = await loadAccounts();
  const context = loadContext();
  const limits = await loadLimits();
  const limiter = new Limiter(limits);
  const outputs = await loadStackOutputs();

  const includeRegion = region;
  const includeAccountKey = accountKey;
  let includeAccountId: string | undefined;
  if (includeAccountKey) {
    includeAccountId = getAccountId(accounts, includeAccountKey);
    if (!includeAccountId) {
      throw new Error(`Cannot find account ${includeAccountKey}`);
    }
  }
  context.centralOperationsAccount = acceleratorConfig.getMandatoryAccountKey('central-operations');
  context.masterAccount = acceleratorConfig.getMandatoryAccountKey('master');

  const accountStacks = new AccountStacks({
    phase: phase.name,
    accounts,
    context,
    useTempOutputDir,
  });

  // Making sure CDK knows all stacks in all regions and accounts
  // This makes sure CDK removes stacks that would otherwise not get deleted
  for (const account of accounts) {
    for (const supportedRegion of acceleratorConfig['global-options']['supported-regions']) {
      accountStacks.tryGetOrCreateAccountStack(account.key, supportedRegion);
    }
  }

  const runner = await phase.runner();
  await runner({
    acceleratorConfig,
    accountStacks,
    accounts,
    context,
    limiter,
    outputs,
  });

  const apps = accountStacks.apps.filter(app => {
    if (includeAccountId && includeAccountId !== app.accountId) {
      console.log(`Skipping app deployment for account ${app.accountKey} and region ${app.region}`);
      return false;
    }
    if (includeRegion && includeRegion !== app.region) {
      console.log(`Skipping app deployment for account ${app.accountKey} and region ${app.region}`);
      return false;
    }
    return true;
  });
  return apps;
}
