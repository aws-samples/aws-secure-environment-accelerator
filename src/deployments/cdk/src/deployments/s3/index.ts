export * from './log-archive-read-access';
