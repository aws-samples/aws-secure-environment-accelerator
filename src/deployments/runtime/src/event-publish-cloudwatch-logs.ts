/**
 *  Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { Context } from 'aws-lambda';
import { throttlingBackOff } from '@aws-accelerator/custom-resource-cfn-utils';
import * as AWS from 'aws-sdk';
import { v4 as uuidv4 } from 'uuid';

const logGroupName = process.env.LOG_GROUP_NAME || '';

const cloudWatchLogs = new AWS.CloudWatchLogs();

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const handler = async (input: any, context?: Context): Promise<void> => {
  console.log('EventBridge CloudWatch Logs Publisher ....');
  console.log(JSON.stringify(input, null, 2));

  console.log(`Publishing to ${logGroupName}`);
  if (logGroupName) {
    const logStreamName = `${new Date().toISOString().slice(0, 10)}-${uuidv4()}`;

    await throttlingBackOff(() =>
      cloudWatchLogs
        .createLogStream({
          logGroupName,
          logStreamName,
        })
        .promise(),
    );

    await throttlingBackOff(() =>
      cloudWatchLogs
        .putLogEvents({
          logGroupName,
          logStreamName,
          logEvents: [
            {
              timestamp: Date.now(),
              message: JSON.stringify(input),
            },
          ],
        })
        .promise(),
    );
  } else {
    console.log(`LogGroupName not specified`);
  }
};
