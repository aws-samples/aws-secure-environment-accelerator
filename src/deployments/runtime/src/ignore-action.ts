// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const handler = async (input: any): Promise<void> => {
  console.log('Ignoring Action input ....');
  console.log(JSON.stringify(input, null, 2));
};
