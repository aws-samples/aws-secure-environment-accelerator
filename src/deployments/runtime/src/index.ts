/**
 *  Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

export { handler as createSnsPublishToCentralRegion } from './sns-publish-central-region';
export { handler as createIgnoreAction } from './ignore-action';
export { handler as albIpMonitor } from './alb-to-alb-target/alb-ip-monitor';
export { handler as albTargetRecordMonitor } from './alb-to-alb-target/alb-target-record-monitor';
export { handler as firehoseCustomPrefix } from './firehose-custom-prefix/process-record';
export { handler as eventToCWLPublisher } from './event-publish-cloudwatch-logs';
export { handler as metadataCollection } from './metadata-collection';
import * as ouValidationEvents from './ou-validation-events';
export { ouValidationEvents };
