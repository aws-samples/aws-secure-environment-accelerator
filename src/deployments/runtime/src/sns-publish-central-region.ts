import { SNS } from '@aws-accelerator/common/src/aws/sns';
import { SNSEvent, Context } from 'aws-lambda';

const logCentralRegion = process.env.CENTRAL_LOG_SERVICES_REGION!;
const sns = new SNS(undefined, logCentralRegion);

export const handler = async (input: SNSEvent, context: Context): Promise<void> => {
  console.log('Verifying Account Creation status ....');
  console.log(JSON.stringify(input, null, 2));
  const snsNotificationConfig = input.Records[0].Sns;
  const topicArn = snsNotificationConfig.TopicArn;
  const topicName = topicArn.split(':').pop();
  const accountId = context.invokedFunctionArn.split(':')[4];
  await sns.publish({
    Message: snsNotificationConfig.Message,
    Subject: snsNotificationConfig.Subject,
    TopicArn: `arn:aws:sns:${logCentralRegion}:${accountId}:${topicName}`,
  });
};
