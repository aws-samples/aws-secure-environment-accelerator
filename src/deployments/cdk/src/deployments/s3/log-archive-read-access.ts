import * as s3 from '@aws-cdk/aws-s3';
import * as iam from '@aws-cdk/aws-iam';
import { AccountStacks } from '../../common/account-stacks';
import { AcceleratorConfig } from '@aws-accelerator/common-config/src';
import { Account, getAccountId } from '@aws-accelerator/common-outputs/src/accounts';
import { S3UpdateLogArchivePolicy } from '@aws-accelerator/custom-resource-s3-update-logarchive-policy';

export interface LogArchiveReadAccessProps {
  accountStacks: AccountStacks;
  accounts: Account[];
  logBucket: s3.IBucket;
  config: AcceleratorConfig;
  acceleratorPrefix: string;
}

export async function logArchiveReadOnlyAccess(props: LogArchiveReadAccessProps) {
  const { accountStacks, accounts, logBucket, config, acceleratorPrefix } = props;
  const logArchiveAccountKey = config['global-options']['central-log-services'].account;
  const logArchiveStack = accountStacks.getOrCreateAccountStack(logArchiveAccountKey);
  const logArchiveReadOnlyRoles = [];

  // Update Log Archive Bucket and KMS Key policies for roles with ssm-log-archive-read-only-access
  for (const { accountKey, iam: iamConfig } of config.getIamConfigs()) {
    const accountId = getAccountId(accounts, accountKey);
    const roles = iamConfig.roles || [];
    for (const role of roles) {
      if (role['ssm-log-archive-read-only-access']) {
        logArchiveReadOnlyRoles.push(`arn:aws:iam::${accountId}:role/${role.role}`);
      }
    }
  }

  const LogBucketPolicy = new S3UpdateLogArchivePolicy(logArchiveStack, 'UpdateLogArchivePolicy', {
    roles: logArchiveReadOnlyRoles,
    logBucket,
    acceleratorPrefix,
  });
}
