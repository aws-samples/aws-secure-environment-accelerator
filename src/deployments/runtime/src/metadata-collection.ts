/* eslint-disable  @typescript-eslint/no-explicit-any */

import { CodeCommit } from '@aws-accelerator/common/src/aws/codecommit';
import { DynamoDB } from '@aws-accelerator/common/src/aws/dynamodb';
import { Organizations } from '@aws-accelerator/common/src/aws/organizations';
import { S3 } from '@aws-accelerator/common/src/aws/s3';
import { SecretsManager } from '@aws-accelerator/common/src/aws/secrets-manager';
import { SSM } from '@aws-accelerator/common/src/aws/ssm';
import { StepFunctions } from '@aws-accelerator/common/src/aws/stepfunctions';
import { STS } from '@aws-accelerator/common/src/aws/sts';

export const handler = async (_event: any, _context: any) => {
  console.log(_event);

  const accleratorPrefix = process.env.ACCELERATOR_PREFIX!;
  const centralBucketName = process.env.CENTRAL_BUCKET_NAME!;
  const bucketName = process.env.BUCKET_NAME!;
  const configRepositoryName = process.env.CONFIG_REPOSITORY_NAME!;
  const outputsTableName = `${accleratorPrefix}Outputs`;
  const acceleratorVersionSSMPath = '/accelerator/version';
  const acceleratorFirstVersionSSMPath = '/accelerator/first-version';
  const lastSuccessfulCommitIdSecretPath = 'accelerator/config/last-successful-commit';
  const codecommit = new CodeCommit();
  const ddb = new DynamoDB();
  const organizations = new Organizations();
  const secrets = new SecretsManager();
  const ssm = new SSM();
  const s3 = new S3();
  const stepFunctions = new StepFunctions();
  const sts = new STS();
  const stateMachineName = `${accleratorPrefix}MainStateMachine_sm`;

  const callerIdentity = await sts.getCallerIdentity();
  const stateMachineArn = `arn:aws:states:${process.env.AWS_REGION}:${callerIdentity.Account}:stateMachine:${stateMachineName}`;

  console.log('Getting accelerator versions');
  const acceleratorCurrentVersionParameter = await ssm.getParameter(acceleratorVersionSSMPath);
  const acceleratorCurrentVersionValue = JSON.parse(acceleratorCurrentVersionParameter.Parameter?.Value || '{}');
  const acceleratorCurrentVersion = acceleratorCurrentVersionValue.Branch;
  const acceleratorFirstVersion = await ssm.getParameter(acceleratorFirstVersionSSMPath);

  console.log(acceleratorCurrentVersionValue);

  //Get latest successful config
  const lastSuccessfulCommitIdSecret = await secrets.getSecret(lastSuccessfulCommitIdSecretPath);
  const lastSuccessfulCommitIdObj = JSON.parse(lastSuccessfulCommitIdSecret.SecretString!);
  const lastSuccessfulCommitId = lastSuccessfulCommitIdObj.configCommitId;
  const rawASEAConfigFile = await codecommit.getFile(configRepositoryName, 'raw/config.json', lastSuccessfulCommitId);
  const aseaConfig = rawASEAConfigFile.fileContent.toString();
  const aseaConfigObj = JSON.parse(aseaConfig);

  //Retrieve Log Bucket and AES Log bucket info
  const logArchiveAccountName = aseaConfigObj['global-options']['central-log-services']['account'];
  const logArchiveRegion = aseaConfigObj['global-options']['central-log-services']['region'];
  const ddbOutputId = `${logArchiveAccountName}-${logArchiveRegion}-0`;
  const outputsItem = await ddb.documentClient.get({ TableName: outputsTableName, Key: { id: ddbOutputId } }).promise();
  const outputs: any[] = JSON.parse(outputsItem?.Item?.outputValue);

  const aesLogBucketOutput = outputs.find((output: any) => {
    return output.outputKey.includes('AesLogBucketOutput');
  });
  const aesLogBucketInfo = JSON.parse(aesLogBucketOutput.outputValue);

  const logBucketOutput = outputs.find((output: any) => {
    return output.outputKey.includes('LogBucketOutput');
  });
  const logBucketInfo = JSON.parse(logBucketOutput.outputValue);

  // Get step function info
  const executions = await stepFunctions.listExecutions({ stateMachineArn, statusFilter: 'SUCCEEDED' });
  const latestSuccessfulExecution = executions.reduce((latest: any, execution: any) => {
    if (!latest) {
      return execution;
    }
    const latestStartDate = new Date(latest.startDate);
    const currentStartDate = new Date(execution.startDate);
    if (latestStartDate >= currentStartDate) {
      return latest;
    }
    return execution;
  }, {});

  // Retrieve Organization Information
  const organizationInfo = await organizations.describeOrganization();
  const organizationId = organizationInfo?.Id;
  const accounts = await organizations.listAccounts();
  const ous = await organizations.listOrganizationalUnits();

  const ousWithPath: any = [];
  for (const ou of ous) {
    const path = await organizations.getOrganizationalUnitWithPath(ou.Id!);
    ousWithPath.push({
      ...ou,
      Path: path.Path,
    });
  }

  //Retrieve Account Information
  const accountsWithOuPath: any = [];
  for (const account of accounts) {
    const accountOuParent = await organizations.listParents(account.Id!);
    const ou = ousWithPath.find((ou: any) => {
      return accountOuParent[0].Id === ou.Id;
    });
    accountsWithOuPath.push({
      ...account,
      Path: ou.Path,
    });
  }

  const metadata = {
    acceleratorFirstVersion: acceleratorFirstVersion.Parameter?.Value,
    acceleratorCurrentVersion,
    lastSuccessfulCommitId,
    latestSuccessfulExecution,
    logBucket: logBucketInfo,
    aesLogBucket: aesLogBucketInfo,
    organizationId,
    accounts: accountsWithOuPath,
    ous: ousWithPath,
  };

  const centralBucketItems = await s3.listBucket(centralBucketName);
  const metadataBucketItems = await s3.listBucket(bucketName);
  const centralBucketItemNames = centralBucketItems.map(item => item.Key);

  const itemsToDelete = metadataBucketItems.filter(item => {
    const splitKey = item.Key?.split('/') || [];
    splitKey.shift();
    const key = splitKey.join('/');
    return !centralBucketItemNames.includes(key);
  });

  const keysToDelete = itemsToDelete.map(item => {
    return { Key: item.Key! };
  });

  if (keysToDelete.length > 0) {
    await s3.deleteObjects({ Bucket: bucketName, Delete: { Objects: keysToDelete } });
  }
  //Delete Utens
  for (const item of centralBucketItems) {
    if (!item.Key?.startsWith('certs/')) {
      await s3.copyObjectWithACL(centralBucketName, item.Key!, bucketName, 'config', 'bucket-owner-full-control');
    }
  }

  console.log('writing metadata and config to bucket.');

  await s3.putObject({
    Bucket: bucketName,
    Key: 'metadata.json',
    Body: JSON.stringify(metadata, null, 4),
    ACL: 'bucket-owner-full-control',
  });
  await s3.putObject({
    Bucket: bucketName,
    Key: 'config/config.json',
    Body: aseaConfig,
    ACL: 'bucket-owner-full-control',
  });
};
