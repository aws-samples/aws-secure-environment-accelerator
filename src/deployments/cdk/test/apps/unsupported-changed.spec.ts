/**
 *  Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import 'jest';
import * as fs from 'fs';
import * as path from 'path';
import * as cdk from '@aws-cdk/core';
import * as cfnspec from '@aws-cdk/cfnspec';
import { S3 } from '@aws-accelerator/common/src/aws/s3';
import { DynamoDB } from '@aws-accelerator/common/src/aws/dynamodb';
import { STS } from '@aws-accelerator/common/src/aws/sts';
import { resourcesToList, ResourceWithLogicalId, ResourceProperties } from '../jest';
import { deployPhases } from './unsupported-changes.mocks';
import { phases } from '../../src/app';

/**
 * List of resource types that need to be tested for immutability.
 */
const resourceTypeNames = [
  'AWS::Budgets::Budget',
  'AWS::DirectoryService::MicrosoftAD',
  'AWS::EC2::Instance',
  'AWS::EC2::TransitGateway',
  'AWS::ElasticLoadBalancingV2::LoadBalancer',
  'AWS::S3::Bucket',
  'AWS::SecretsManager::Secret',
  'AWS::SecretsManager::ResourcePolicy',
];

type PropertySelector = (properties: ResourceProperties) => ResourceProperties;

/**
 * Return a copy of the given properties with only the given property names.
 */
function selectProperties(properties: ResourceProperties, propertyNames: string[]): ResourceProperties {
  return propertyNames.reduce((result, propertyName) => ({ ...result, [propertyName]: properties[propertyName] }), {});
}

/**
 * Create property selector that selects immutable properties from the given properties.
 */
function createImmutablePropertySelector(type: cfnspec.schema.ResourceType): PropertySelector {
  // Find all names of immutable properties
  const immutablePropertyNames: string[] = Object.entries(type.Properties || {})
    .filter(([_, propertySpec]) => propertySpec.UpdateType === cfnspec.schema.UpdateType.Immutable)
    .map(([propertyName, _]) => propertyName);

  // Create a closure that selects the immutable properties from the given properties
  return properties => selectProperties(properties, immutablePropertyNames);
}

const stackResources: { [stackName: string]: ResourceWithLogicalId[] } = {};

beforeAll(async () => {
  // Mock STS and S3 as the IAM logic is currently contacting S3 using the SDK
  // @ts-ignore
  jest.spyOn(STS.prototype, 'getCredentialsForAccountAndRole').mockImplementation(() => Promise.resolve(undefined));
  jest.spyOn(S3.prototype, 'getObjectBodyAsString').mockImplementation(() =>
    Promise.resolve(
      JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Effect: 'Allow',
            Action: '*',
            Resource: '*',
          },
        ],
      }),
    ),
  );
  // Mock DynamoDB as the VPC and Subnet CIDR retrival using dynamodb.scan using the SDK
  jest.spyOn(DynamoDB.prototype, 'scan').mockImplementation(() => Promise.resolve([]));

  // Deploy all phases that are defined in src/app.ts
  for await (const app of deployPhases(phases)) {
    const assembly = app.synth();
    const stacks = app.node.children.filter(cdk.Stack.isStack);
    for (const stack of stacks) {
      const artifact = assembly.getStackArtifact(stack.artifactId);
      const template = artifact.template;
      // eslint-disable-next-line deprecation/deprecation
      stackResources[stack.node.uniqueId] = resourcesToList(template.Resources);

      // Render all nested stacks
      // See https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/assert/lib/synth-utils.ts#L52
      const nestedStacks = stack.node.findAll().filter(cdk.NestedStack.isNestedStack);
      for (const nestedStack of nestedStacks) {
        const nestedTemplateFile = path.join(assembly.directory, nestedStack.templateFile);
        const nestedTemplate = JSON.parse(fs.readFileSync(nestedTemplateFile).toString('utf-8'));
        // eslint-disable-next-line deprecation/deprecation
        stackResources[nestedStack.node.uniqueId] = resourcesToList(nestedTemplate.Resources);
      }
    }
  }
});

test.each(resourceTypeNames)('there should not be any unsupported resource changes for %s', resourceTypeName => {
  const resourceSpec = cfnspec.resourceSpecification(resourceTypeName);
  const selectProperties = createImmutablePropertySelector(resourceSpec);

  for (const [stackName, resources] of Object.entries(stackResources)) {
    // Find all resources with the given type
    const resourcesOfType = resources.filter(r => r.Type === resourceTypeName);

    // Select the relevant properties
    const expected = resourcesOfType.map(resource => ({
      LogicalId: resource.LogicalId,
      Properties: selectProperties(resource.Properties),
    }));

    // Compare the relevant properties to the snapshot
    expect(expected).toMatchSnapshot(stackName);
  }
});

// test('templates should stay exactly the same', () => {
//   for (const [stackName, resources] of Object.entries(stackResources)) {
//     // Compare the relevant properties to the snapshot
//     expect(resources).toMatchSnapshot(stackName);
//   }
// });
