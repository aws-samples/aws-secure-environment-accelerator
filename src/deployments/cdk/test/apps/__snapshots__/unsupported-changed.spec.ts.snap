// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`there should not be any unsupported resource changes for AWS::Budgets::Budget: FunAcctPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::Budgets::Budget: LogArchivePhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::Budgets::Budget: MasterPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::Budgets::Budget: Mydevacct1Phase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::Budgets::Budget: OperationsPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::Budgets::Budget: PerimeterPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::Budgets::Budget: SecurityPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::Budgets::Budget: SharedNetworkPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::Budgets::Budget: SharedServicesPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::DirectoryService::MicrosoftAD: FunAcctPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::DirectoryService::MicrosoftAD: LogArchivePhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::DirectoryService::MicrosoftAD: MasterPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::DirectoryService::MicrosoftAD: Mydevacct1Phase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::DirectoryService::MicrosoftAD: OperationsPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::DirectoryService::MicrosoftAD: PerimeterPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::DirectoryService::MicrosoftAD: SecurityPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::DirectoryService::MicrosoftAD: SharedNetworkPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::DirectoryService::MicrosoftAD: SharedServicesPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::EC2::Instance: FunAcctPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::EC2::Instance: LogArchivePhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::EC2::Instance: MasterPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::EC2::Instance: Mydevacct1Phase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::EC2::Instance: OperationsPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::EC2::Instance: PerimeterPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::EC2::Instance: SecurityPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::EC2::Instance: SharedNetworkPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::EC2::Instance: SharedServicesPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::EC2::TransitGateway: FunAcctPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::EC2::TransitGateway: LogArchivePhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::EC2::TransitGateway: MasterPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::EC2::TransitGateway: Mydevacct1Phase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::EC2::TransitGateway: OperationsPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::EC2::TransitGateway: PerimeterPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::EC2::TransitGateway: SecurityPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::EC2::TransitGateway: SharedNetworkPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::EC2::TransitGateway: SharedServicesPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::ElasticLoadBalancingV2::LoadBalancer: FunAcctPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::ElasticLoadBalancingV2::LoadBalancer: LogArchivePhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::ElasticLoadBalancingV2::LoadBalancer: MasterPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::ElasticLoadBalancingV2::LoadBalancer: Mydevacct1Phase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::ElasticLoadBalancingV2::LoadBalancer: OperationsPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::ElasticLoadBalancingV2::LoadBalancer: PerimeterPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::ElasticLoadBalancingV2::LoadBalancer: SecurityPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::ElasticLoadBalancingV2::LoadBalancer: SharedNetworkPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::ElasticLoadBalancingV2::LoadBalancer: SharedServicesPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::S3::Bucket: FunAcctPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::S3::Bucket: LogArchivePhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::S3::Bucket: MasterPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::S3::Bucket: Mydevacct1Phase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::S3::Bucket: OperationsPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::S3::Bucket: PerimeterPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::S3::Bucket: SecurityPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::S3::Bucket: SharedNetworkPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::S3::Bucket: SharedServicesPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::SecretsManager::ResourcePolicy: FunAcctPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::SecretsManager::ResourcePolicy: LogArchivePhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::SecretsManager::ResourcePolicy: MasterPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::SecretsManager::ResourcePolicy: Mydevacct1Phase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::SecretsManager::ResourcePolicy: OperationsPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::SecretsManager::ResourcePolicy: PerimeterPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::SecretsManager::ResourcePolicy: SecurityPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::SecretsManager::ResourcePolicy: SharedNetworkPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::SecretsManager::ResourcePolicy: SharedServicesPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::SecretsManager::Secret: FunAcctPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::SecretsManager::Secret: LogArchivePhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::SecretsManager::Secret: MasterPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::SecretsManager::Secret: Mydevacct1Phase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::SecretsManager::Secret: OperationsPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::SecretsManager::Secret: PerimeterPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::SecretsManager::Secret: SecurityPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::SecretsManager::Secret: SharedNetworkPhase1 1`] = `Array []`;

exports[`there should not be any unsupported resource changes for AWS::SecretsManager::Secret: SharedServicesPhase1 1`] = `Array []`;
